feat!: use inspector for time profiles

aabmass opened the pull request:
For time profiling, replaces the native C++ addon with calls to [Node's inspector module](https://nodejs.org/docs/latest-v14.x/api/inspector.html#inspector_cpu_profiler). This uses the same `V8::CpuProfiler` under the hood but communicates over the chrome devtools protocol. More context: https://github.com/google/pprof-nodejs/issues/170#issuecomment-1176332831

Once the change is also made for heap profilers, this will obsolete the C++ code and we will no longer need any prebuilt binaries.

## Breaking changes 📢 
- `name` and `lineNumbers` options removed from `TimeProfilerOptions`
- `time.start()`signature changed; it is now async and the returned stop function is also async.
- line numbers in profiles are now reported as [`CpuProfilingMode::kLeafNodeLineNumbers`](https://source.chromium.org/chromium/chromium/src/+/main:v8/include/v8-profiler.h;l=244;drc=27e39fa1dd71076618c358639ed8a327bc3873c4;bpv=1;bpt=1) instead of [`CpuProfilingMode::kCallerLineNumbers`](https://source.chromium.org/chromium/chromium/src/+/main:v8/include/v8-profiler.h;l=247;drc=27e39fa1dd71076618c358639ed8a327bc3873c4;bpv=1;bpt=1)
- requiring the module no longer works and will be removed https://github.com/google/pprof-nodejs/issues/234

## Testing
I wrote a test script to generate some profiles and check for any differences. The flame graphs look the same:
![image](https://user-images.githubusercontent.com/1510004/197109775-59fcc295-b285-4efe-996a-6867ef4d5322.png)

Main differences I can see in the peek view of the profiles:
![image](https://user-images.githubusercontent.com/1510004/197109651-78cd8b92-4900-4763-970a-be883c419299.png)

aabmass committed:
feat!: use inspector for time profiles

aabmass commented on ts/src/index.ts:
```diff
@@ -37,12 +37,16 @@ export const heap = {
+      // The process is going to terminate imminently. All work here needs to
+      // be synchronous.
+
+      // TODO: this code no longer works because stop() cannot be run synchronously. Maybe
+      // beforeExit event would be a decent middleground, or can run this in a signal handler.
```
Any thoughts on this?

nolanmar511 replied:
This is a breaking change, but the profiling agent for Node.js doesn't use this, so the breaking change seems fine to me.

I don't think it will impact integration tests ([busybench.ts](https://github.com/google/pprof-nodejs/blob/5ae0e9b2581cf62d051ef151925d0b1eb79e9d9c/system-test/busybench/src/busybench.ts) and [busybench.js](https://github.com/google/pprof-nodejs/blob/main/system-test/busybench-js/src/busybench.js) both invoke the profiler from their code, so should still work), so (beyond perhaps noting this change in release note), eliminating this should be too much effort.

aabmass replied:
Made a follow up issue to remove the code and docs https://github.com/google/pprof-nodejs/issues/234

aabmass commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+      if (err !== null) {
+        reject(err);
+      }
+      session.post('Profiler.start', err => {
```
There's no option to pass runName (title) in this API. Is this problematic for overlapping profiles?

- https://nodejs.org/api/inspector.html#cpu-profiler_1
- calls this dev tools protocol method https://chromedevtools.github.io/devtools-protocol/v8/Profiler/#method-enable

nolanmar511 replied:
Overlapping profiles will never happen if a user only has the Node.js profiling agent running (since the agent will collect a profile before asking the backend if it can collect another). 

This difference is likely fine.

aabmass replied:
For now I'm just removing the name param altogether. If we get FRs in the future, we can add this session mechanism.

aabmass commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+    const parsedScriptId = parseInt(scriptId);
+    return {
+      name: functionName,
+      scriptName: url,
```
Url will have a `file:/` scheme, which differs from the C++ impl

nolanmar511 replied:
This is probably a breaking change from the perspective of profile consumers (Cloud Profiler's UI doesn't have a way to ignore file names when doing profile comparison -- so the flame graph will fork when merging profiles across versions & doing diff). But, other than that, I don't see this posing any major issues.

aabmass replied:
Added code to strip this prefix when present anyway. I imagine the URI is mainly intended for browser scripts but for node will usually be `file:/` and repetitive.

aabmass commented on ts/src/time-profiler.ts:
```diff
@@ -49,40 +49,36 @@ export interface TimeProfilerOptions {
 }
 
-export function start(
+export async function start(
```
This now has to be async which is another breaking change.

nolanmar511 replied:
There's a chance (depending on underlying implementation) this will result in a fix for https://github.com/googleapis/cloud-profiler-nodejs/issues/683 (event loop blocked at start of profile collection). 

This would be a good change if it does that (though needing to update agent code will be a bummer)

aabmass replied:
Great! I will assume this fixes that issue and close it once this change gets pull into that repo.

nolanmar511 commented:
On the off-by-one line numbers & differing file names, the existing system tests checking that mapping of javascript to typescript (uses busybench.ts; [here](https://github.com/google/pprof-nodejs/blob/5ae0e9b2581cf62d051ef151925d0b1eb79e9d9c/system-test/test.sh#L67-L70)) might be helpful. 

As long as the mapping of javascript => typescript works, any slight changes in file names and line numbers seem fine to me.

punya commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright 2018 Google Inc. All Rights Reserved.
```
Nit: 2022

punya commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+  return new Promise<void>((resolve, reject) => {
+    session.post('Profiler.enable', err => {
+      if (err !== null) {
+        reject(err);
```
```suggestion
        reject(err);
        return;
```

punya commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+      }
+      session.post('Profiler.start', err => {
+        if (err !== null) {
+          reject(err);
```
```suggestion
          reject(err);
          return;
```

punya commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+  return new Promise<TimeProfile>((resolve, reject) => {
+    session.post('Profiler.stop', (err, {profile}) => {
+      if (err !== null) {
+        reject(err);
```
```suggestion
        reject(err);
        return;
```

punya commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,113 @@
+      {interval: intervalMicros},
+      err => {
+        if (err !== null) {
+          reject(err);
```
```suggestion
          reject(err);
          return;
```

punya commented:
Node 12 is EOL so I think we should be okay with dropping that from CI.

aabmass committed:
Apply suggestions from code review

Co-authored-by: Punya Biswal <punya@google.com>

aabmass committed:
Merge branch 'main' into use-inspector-cpu

aabmass committed:
Small fixes

aabmass committed:
More fixes

aabmass committed:
Merge branch 'main' into use-inspector-cpu

aabmass committed:
Strip file:/ prefixes from scriptNames and fixed system tests

The inspector approach uses
[`CpuProfilingMode::kLeafNodeLineNumbers`](https://source.chromium.org/chromium/chromium/src/+/main:v8/include/v8-profiler.h;l=244;drc=27e39fa1dd71076618c358639ed8a327bc3873c4)
under the hood which changes the line numbering scheme. I updated the
tests to expect the new line number.

aabmass committed:
Remove unnused options since we are already making breaking changes for v4

aabmass committed:
Add todo for --require

aabmass committed:
Undo accidental edits to system test

aabmass committed:
Add explicit signatures to modified functions to make breaking changes more clear going forward

weyert commented:
What's the status of this PR? Really interested in trying out Cloud Profiler with Node LTS  :D

psx95 commented on ts/src/time-profiler-inspector.ts:
```diff
@@ -0,0 +1,117 @@
+import * as inspector from 'node:inspector';
+
+const session = new inspector.Session();
+session.connect();
```
Shouldn't this have a corresponding `disconnect` call ?  Would it make sense to move it inside `startProfiling` ?

aabmass replied:
Ya probably should do this 👍 see this comment https://github.com/google/pprof-nodejs/pull/227#discussion_r1007443301

Only thing I can think is that it may be less efficient, but looking at the V8 code I'm not sure if that's true. Would probably need to test it.

aabmass commented:
> What's the status of this PR? Really interested in trying out Cloud Profiler with Node LTS :D

@weyert We want to verify that HeapProfiler through inspector is working correctly too. Then we can remove all the native code and make a release

weyert commented:
> @weyert We want to verify that HeapProfiler through inspector is working correctly too. Then we can remove all the native code and make a release

Thanks for the update, have to admit not fully following what you are saying but I wish you good luck finding out :D

psx95 commented:
> > @weyert We want to verify that HeapProfiler through inspector is working correctly too. Then we can remove all the native code and make a release
> 
> Thanks for the update, have to admit not fully following what you are saying but I wish you good luck finding out :D

So, the in-built node profiler also has [heap profiler](https://nodejs.org/docs/latest-v14.x/api/inspector.html#inspector_heap_profiler) functionality. We need to verify if that can be used to replace heap profiling here. If it works, we can make a new release which uses node's in-built profilers for profiling both - CPU & Memory instead of using native code which we currently do.

weyert commented:
Any news on the support for Node 16/18?

aabmass commented:
Closing as very stale
